File: Interview Bit/Binary Search/search-for-a-range.cpp
int first(const vector<int> &A,int x,int low,int high,int n){
	
	if(high >= low){
		
		int mid = low + (high - low)/2;
		
		if( (mid == 0 || x > A[mid-1]) && (A[mid] == x)){
			return mid;
		}else if(x > A[mid]){
			return first(A,x,mid + 1,high,n);
		}else{
			return first(A,x,low,mid - 1,n);
		}
	}
	
	return -1;

}


int last(const vector<int> &A,int x,int low,int high,int n){
	
	if(high >= low){
		
		int mid = low + (high - low)/2;
		
		if((mid == n-1 || x < A[mid+1]) && A[mid] == x){
			return  mid;
		}else if(x < A[mid]){
			return last(A,x,low,mid - 1,n);
		}else{
			return last(A,x,mid+1,high,n);
		}
	}
	
	return -1;
}

File: Interview Bit/Binary Search/search-for-a-range.py
    def searchRange(self, A, B):
            first=-1
            start=0
            end=len(A)-1
    
            while(start<=end):
                mid = int(start + (end - start)//2)
                if (mid == 0 or B > A[mid - 1]) and (A[mid] == B):
                    first = mid
                    break
                elif B > A[mid]:
                    start = mid + 1
                else:
                    end = mid - 1
            
            if first == -1:
                return [-1,-1]
            last=-1
            start=0
            end=len(A)-1
            while (start<=end):
                mid = int(start + (end - start)//2)
                if (mid == len(A)-1 or B < A[mid + 1] ) and (A[mid] == B):
                    last = mid
                    break;
                elif B < A[mid]:
                    end = mid - 1
                else:
                    start = mid + 1
            
            return [first,last]
        